rfcs/0156-no-direct-pushes.md

```diff
@@ -0,0 +1,173 @@
+---
+feature: no-direct-pushes
+start-date: 2023-07-21
+author: Silvan Mosberger
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Require pull requests for all Nixpkgs commits.
+
+# Motivation
+[motivation]: #motivation
+
+There are currently 197 people with commit access to Nixpkgs, all of whom can push directly to any branch without a pull request.
+Such pushes generally do not notify anybody and do not trigger CI.
+
+This makes such commits susceptible to:
+- Be anonymous
+- Include malicious code
+- Be broken
+- Have poor code quality
```

Comment:
All of these would also be addressed by "Require signed commits || Require a pull request before merging".

Reply:
As long as we use CI tied to PRs that depends on master evaluating with only specific deviations from cleanly, «be broken» does get easier to handle without direct pushes…

Reply:
I'm pretty sure signed commits doesn't make it harder to push malicious, broken or poor-quality code, why would it? The only thing it gets you is that commits can't be anonymous anymore.

Reply:
I see, yeah I didn't consider such an attack scenario here. I'll definitely add this appropriately into the Alternatives and Future Work sections.

Reply:
Require signing does not help at all. Right now nixpkgs has 10k forks and 5k+ contributors. Require signing does not offer anything, as it would be just another contributor, another ssh key, another github account that pushes a PR. There's no value if you don't use a whitelist of allowed contributors and a network of trust or something like that.
Also, changes made through the github web UI would be signed by the github key... so there's no benefit there either.

Or am I missing something?

Reply:
> If I am correct it's also not possible anymore to make changes via the API as it's lacking the signing key. Of course, our contributors can always still push something broken themselves.

I believe changes "made from github" are intrinsically signed.

See this commit made from the GitHub UI:

 - https://github.com/samueldr-wip/mobile-nixos-wip/commit/7ffd2b4e07ef7fb123e39f5a7a6f122b23d718d9

I am assuming, but my assumption is that changes made through the API would also be "Verified" using "GitHub’s verified signature".

Reply:
That's the exact opposite of what I thought the API behaviour was, thanks for demonstrating!

Reply:
> On the contrary. The crux of the issue is increasing our level of dependence on git-hub-specific features.

This can and does apply to any "forge" or any process for development.

It's not "must be github produced pull requests", but "tangible item that allows tracking the origin". If Nixpkgs had a mailing-lists development process, it would be "any change must be sent as a patch set / pull request", and thus provide the same tangible "item".

The principle behind the change does not entrenches into any specific tooling, it makes a pretty much already universal *development process* mandatory.

> Neither does opening a PR and merging it.

This leaves a common tangible "item" to discuss about the bad contribution. Thus, the overall already near-universal *development process* is defined.

@infinisil do tell if you think I've misunderstood something here.

Reply:
Yeah you're on point @samueldr. And while I do like the thought of switching off GitHub and signing commits, this is not the place to discuss it.

Reply:
> The crux of the issue is increasing our level of dependence on git-hub-specific features.

I'd also add to this that while GitHub has a UI for this, this is not GitHub specific. If it's later decided to host our own copy of the canonical Nix repository somewhere, you could still get the same feature this RFC asks for with a remote commit hook.

Reply:
> It's not "must be github produced pull requests", but "tangible item that allows tracking the origin".

The RFC does not say "tangible item that allows tracking the origin".

It says:

https://github.com/tweag/rfcs/blob/709c8979ece291291ff12da8e206cb212a14652e/rfcs/0156-no-direct-pushes.md?plain=1#L43

> this is not GitHub specific

The text above specifically references git-hub.

Reply:
> The text above specifically references git-hub.

Aw come on, this is needlessly pedantic quibbling. Yes this is a GitHub feature, and therefore the RFC text mentions (and should mention) GitHub. No this is not a GitHub *specific* feature, as most other git forges offer a 1:1 equivalent.